pkg/ipam/strictvl3ipam/server.go

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2024 Cisco and its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package strictvl3ipam provides a networkservice.NetworkService Server chain element that resets IP context configuration out of the settings scope
+package strictvl3ipam
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/ippool"
+)
+
+type strictVl3IPAMServer struct {
+	ipPool *ippool.IPPool
+}
+
+// NewServer - returns a new ipam networkservice.NetworkServiceServer that validates the incoming IP context parameters and resets them based on the validation result.
+func NewServer(ctx context.Context, vl3IPAMServer networkservice.NetworkServiceServer, ipPool *ippool.IPPool) networkservice.NetworkServiceServer {
+	return next.NewNetworkServiceServer(
+		&strictVl3IPAMServer{ipPool: ipPool},
+		vl3IPAMServer,
+	)
+}
+
+func (s *strictVl3IPAMServer) areAddressesValid(addresses []string) bool {
+	s.ipPool.Lock()
+	defer s.ipPool.Unlock()
+
+	if len(addresses) == 0 {
+		return true
+	}
+
+	for _, addr := range addresses {
+		if s.ipPool.ContainsNetString(addr) {
+			return true
+		}
+	}
+	return false
+}
+
+func (s *strictVl3IPAMServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	srcAddrs := request.GetConnection().GetContext().GetIpContext().GetSrcIpAddrs()
+	dstAddrs := request.GetConnection().GetContext().GetIpContext().GetDstIpAddrs()
+
+	if !s.areAddressesValid(srcAddrs) || !s.areAddressesValid(dstAddrs) {
+		request.Connection.Context.IpContext = &networkservice.IPContext{}
```

Comment:
This is a wrong way. 
A vl3 client can come to the vl3 server on the first request, and this request will already contain the src address of the vl3 client.
This way we will delete addresses that we should not have deleted. 
Perhaps we can only work with dst address.
Please run the vl3 example and see what requests vl3-server works with.